### PR TITLE
Get processors and calibrations in the Engine

### PR DESCRIPTION
**Pull request by wcourtney**

### CODE REVIEW

File: cirq/google/engine/engine.py [hunk 1]
```diff
@@ -619,6 +619,103 @@ def remove_job_labels(self, job_resource_name: str, label_keys: List[str]):
             fingerprint = job.get('labelFingerprint', '')
             self._set_job_labels(job_resource_name, new_labels, fingerprint)
 
+    def list_processors(self, project_id: str) -> List[Dict]:
+        """ Returns a list of Processors that the user has visibility to in the
```
Comment (dabacon) [inline cirq/google/engine/engine.py]: Nit, alignment of doc string should be against the """ and should just word wrap.  

Similar in other methods below.

File: cirq/google/engine/engine.py [hunk 2]
```diff
@@ -619,6 +619,103 @@ def remove_job_labels(self, job_resource_name: str, label_keys: List[str]):
...
+            provided project. The names of these processors are used to identify
+            devices when scheduling jobs and gathering calibration metrics.
+
+        Params:
+            project_id: The ID of the Google Cloud project to check, e.g.
+                `my-project-123`
+
+        Returns:
+            A list of dictionaries containing the metadata of each processor.
```
Comment (dabacon) [inline cirq/google/engine/engine.py]: should we doc what is in these dictionaries?

File: cirq/google/engine/engine.py [hunk 3]
```diff
@@ -619,6 +619,103 @@ def remove_job_labels(self, job_resource_name: str, label_keys: List[str]):
...
+        """
+        parent = 'projects/%s' % (project_id)
+        response = self.service.projects().processors().list(
+            parent=parent).execute()
+        return response['processors']
+
+    def get_latest_calibration(self,
+                               processor_name: str) -> Optional['Calibration']:
+        """ Returns metadata about a the latest known calibration run for a
```
Comment (dabacon) [inline cirq/google/engine/engine.py]: "a the" -> "the"

File: cirq/google/engine/engine.py [hunk 4]
```diff
@@ -619,6 +619,103 @@ def remove_job_labels(self, job_resource_name: str, label_keys: List[str]):
...
+        processor, or None if there is no calibration available.
+
+        Params:
+            processor_name: A string of the form
+                `projects/project_id/processors/processor_id`.
+
+        Returns:
+            A dictionary containing the metadata.
```
Comment (dabacon) [inline cirq/google/engine/engine.py]: containing the callibration data?

File: cirq/google/engine/engine.py [hunk 5]
```diff
@@ -619,6 +619,103 @@ def remove_job_labels(self, job_resource_name: str, label_keys: List[str]):
...
+
+
+class Calibration:
+    """A convenience wrapper for calibrations"""
+
+    def __init__(self, calibration: Dict) -> None:
+        self._time = int(calibration['timestampMs'])
+        self._metrics = calibration['metrics']
+
+    def get_timestamp(self) -> int:
```
Comment (dabacon) [inline cirq/google/engine/engine.py]: I'd make this a property and just call it timestamp.  Or alternatively in python form, just make it a attribute and doc that in the doc string.

File: cirq/google/engine/engine.py [hunk 6]
```diff
@@ -619,6 +619,103 @@ def remove_job_labels(self, job_resource_name: str, label_keys: List[str]):
...
+            # (e.g. proto version of each value is {<type>: value}).
+            'values':
+            list(value
+                 for values in metric['values']
+                 for value in list(values.values()))
+        }
+                    for metric in self._metrics
+                    if metric['name'] == name)
+
+    def _values_from_metric(self, metric: Dict) -> List:
```
Comment (dabacon) [inline cirq/google/engine/engine.py]: unused

File: cirq/google/engine/engine.py [hunk 7]
```diff
@@ -619,6 +619,103 @@ def remove_job_labels(self, job_resource_name: str, label_keys: List[str]):
...
+
+        Params:
+            name: the name of a metric referred to in the calibration. Valid
+            names can be found with the get_metric_names() method.
+
+        Returns:
+            A list of dictionaries containing pairs of targets and values for
+            the requested metric.
+        """
+        return list({
```
Comment (dabacon) [inline cirq/google/engine/engine.py]: my preference would be to write this as loops.  when these get multiple lines I find them harder to read.

Comment (wcourtney) [reply cirq/google/engine/engine.py]: Rewritten to use mutation + loops. WDYT?

File: cirq/google/engine/engine.py (hunk 2 above)
Comment (wcourtney) [reply cirq/google/engine/engine.py]: I'm concerned about this getting out of sync, but it does seem useful to get a sense of what to expect so I made a pass at what's returned today. PTAL.

Comment (wcourtney) [reply cirq/google/engine/engine.py]: The service proto isn't published, so removing docs as discussed offline.

**Pull request merged**